src/LanguageServer/Test/IndexParserTests.cs

```diff
@@ -0,0 +1,159 @@
+            _fileSystem.FileExists(testFilePath).Returns(true);
+            SetFileOpen(_fileSystem, testFilePath, _ => throw new FileNotFoundException());
+
+            IIndexParser indexParser = new IndexParser(_symbolIndex, _fileSystem, _pythonLanguageVersion);
+            await indexParser.ParseAsync(testFilePath);
+
+            var symbols = _symbolIndex.WorkspaceSymbols("");
+            symbols.Should().HaveCount(0);
+        }
+
+        [TestMethod, Priority(0)]
+        public void CancellParsingAsync() {
+            const string testFilePath = "C:/bla.py";
+            _fileSystem.FileExists(testFilePath).Returns(true);
+            SetFileOpen(_fileSystem, testFilePath, MakeStream("x = 1"));
+
+            IIndexParser indexParser = new IndexParser(_symbolIndex, _fileSystem, _pythonLanguageVersion);
+            CancellationTokenSource cancellationTokenSource = new CancellationTokenSource();
+            cancellationTokenSource.Cancel();
+
+            Func<Task> parse = async () => {
+                await indexParser.ParseAsync(testFilePath, cancellationTokenSource.Token);
+            };
+            parse.Should().Throw<TaskCanceledException>();
+
+            var symbols = _symbolIndex.WorkspaceSymbols("");
+            symbols.Should().HaveCount(0);
+        }
+
+        private void SetFileOpen(IFileSystem fileSystem, string path, Stream stream) {
+            fileSystem.FileOpen(path, FileMode.Open, FileAccess.Read, FileShare.Read).Returns(stream);
+        }
+
+        [TestMethod, Priority(0)]
+        public void DisposeParserCancelsParsing() {
+            const string testFilePath = "C:/bla.py";
+            _fileSystem.FileExists(testFilePath).Returns(true);
+            ManualResetEventSlim neverSignaledEvent = new ManualResetEventSlim(false);
+            ManualResetEventSlim fileOpenedEvent = new ManualResetEventSlim(false);
+            SetFileOpen(_fileSystem, testFilePath, _ => {
+                fileOpenedEvent.Set();
+                // Wait forever
+                neverSignaledEvent.Wait();
+                throw new InternalTestFailureException("Task should have been cancelled");
+            });
+
+            IIndexParser indexParser = new IndexParser(_symbolIndex, _fileSystem, _pythonLanguageVersion);
+
+            var parseTask = indexParser.ParseAsync(testFilePath);
+            fileOpenedEvent.Wait();
+            indexParser.Dispose();
+            Func<Task> parse = async () => {
+                await parseTask;
+            };
+
+            parse.Should().Throw<TaskCanceledException>();
+
+            var symbols = _symbolIndex.WorkspaceSymbols("");
+            symbols.Should().HaveCount(0);
+        }
+
+        private void SetFileOpen(IFileSystem fileSystem, string path, Func<object, Stream> p) {
+            fileSystem.FileOpen(path, FileMode.Open, FileAccess.Read, FileShare.Read).Returns(p);
+        }
+
+        private Stream MakeStream(string str) {
```

Comment:
This should probably be deleted and instead be `StringReader` everywhere instead.

Reply:
Still applies, you don't need a memory stream for strings.

Reply:
My mistake, `StringReader` is not a `Stream`, since `Stream` is on bytes and `StringReader` is `TextReader`.